golang/cosmos/x/swingset/keeper/keeper.go

```diff
@@ -35,19 +36,71 @@ type Keeper struct {
 
 var _ types.SwingSetKeeper = &Keeper{}
 
-// A prefix of bytes, since KVStores can't handle empty slices as keys.
-var keyPrefix = []byte{':'}
+// Keys are structured as: `${numberOfPathElements}\0${zeroSeparatedPath}`, such
+// as `0\0` for the root element, and `1\0foo` for `foo`, and `3\0foo\0bar\0baz`
+// for `foo.bar.baz`.  The reason for this is space efficiency:
+//   - we can read a node's data with just a single store operation
+//     after mapping the key
+//   - we can scan for all of `foo.bar`'s children by iterating over the prefix
+//    `3\0foo\0bar\0`.
+// We still need to iterate up the tree until we are sure the correct ancestor
+// nodes are present or absent, but we don't need to fetch all an ancestor's
+// keys to do so.
```

Comment:
Need to document the two kinds of keys you're using, and the invariants between the two stores, and really ought to define a "path" here, or point to its definition somewhere else.  I suggest something like the following:

- A "path" is a sequence of zero or more dot-separated nonempty strings of 7-bit non-nul, non-dot ASCII characters. So `""`, `"foo"`, and `"foo.bar.baz"` are paths but `"."`, "foo.", and "fo\0o" are not.
- A storage key for a path is the path prefixed with `:`.
- A path key for a path is the path prefixed with its length (in ASCII digits), separated by nul, followed by the path with dots replaced with nul. So the path key for the empty path is `0\0`.
- Path store entries have just a placeholder value. Path store entries exist if and only if self or some descendant have a non-empty data entry.